### PR TITLE
use gopkg.in/laverya/yaml.v3 instead of gopkg.in/yaml.v2

### PR DESCRIPTION
**Pull request by laverya**
this supports 64 bit floats and longer string lines
see #18 

(`laverya/yaml.v3` instead of `yaml.v3` because the fork supports setting line length, and the original has no releases to pin)

A disadvantage of this upgrade is that there is no longer `UnmarshalStrict` - _all_ `Unmarshal` calls are now strict. **As mentioned later in the thread, this likely constitutes a breaking change.**

### CODE REVIEW

**Comment by laverya** [comment]
I missed the addition of `JSONObjectToYAMLObject`. That's going to be interesting to port to `yaml.v3`, since the type it uses (`yaml.MapSlice`) no longer exists in v3...

**Comment by laverya** [comment]
And so here is a version that keeps `go-yaml/yaml.v2` around for use with `JSONObjectToYAMLObject`. I'm not entirely happy with this, but I'm not sure if there's any other way to move to `go-yaml/yaml.v3`

**Comment by dims** [comment]
cc @deads2k @liggitt

**Comment by liggitt** [comment]
cc @sttts

**Comment by sttts** [comment]
Background: we need JSONObjectToYAMLObject and the (exact) v2 MapSlice type to interface with https://github.com/googleapis/gnostic.

**Comment by sttts** [comment]
> A disadvantage of this upgrade is that there is no longer `UnmarshalStrict` - _all_ `Unmarshal` calls are now strict.

What does this mean in practice? Which existing UX does this break?

**Comment by laverya** [comment]
> > A disadvantage of this upgrade is that there is no longer UnmarshalStrict - all Unmarshal calls are now strict.
> 
> What does this mean in practice? Which existing UX does this break?

It means that the yaml parser will throw an error if duplicate keys are detected. I do not know of any specific use cases that this would break, though - besides the obvious one of "parse whatever junk the users threw at me". **On reflection, this may constitute a breaking API change**.

There _is_ still a difference between `kubernetes-sigs/yaml`'s `Unmarshal` and `UnmarshalStrict` functions, though - namely, `UnmarshalStrict` will throw an error if fields exist in the object being converted that do not exist in the destination object.

> Background: we need JSONObjectToYAMLObject and the (exact) v2 MapSlice type to interface with https://github.com/googleapis/gnostic.

Ouch. That project seems to use MapSlice a lot, and it's never fun to rely on a type/method/etc that gets removed by a major version bump...

**Comment by laverya** [comment]
@sttts 

> > A disadvantage of this upgrade is that there is no longer UnmarshalStrict - all Unmarshal calls are now strict.
> 
> What does this mean in practice? Which existing UX does this break?

It means that the yaml parser will throw an error if duplicate keys are detected. I do not know of any specific use cases that this would break, though - besides the obvious one of "parse whatever junk the users threw at me". **On reflection, this may constitute a breaking API change**.

There _is_ still a difference between `kubernetes-sigs/yaml`'s `Unmarshal` and `UnmarshalStrict` functions, though - namely, `UnmarshalStrict` will throw an error if fields exist in the object being converted that do not exist in the destination object.

> Background: we need JSONObjectToYAMLObject and the (exact) v2 MapSlice type to interface with https://github.com/googleapis/gnostic.

Ouch. That project seems to use MapSlice a lot, and it's never fun to rely on a type/method/etc that gets removed by a major version bump...

File: go.mod [hunk 1]
```diff
@@ -4,5 +4,6 @@ go 1.12
 
 require (
 	github.com/davecgh/go-spew v1.1.1
+	gopkg.in/laverya/yaml.v3 v3.0.0-beta1
```
Comment (neolit123) [inline go.mod]: why include v3 from your fork. what differences it has from upstream?

Comment (laverya) [reply go.mod]: Two major ones - the first is that it has tags
The second is the inclusion of https://github.com/go-yaml/yaml/pull/455 - "Add a function to set the desired line length of an Encoder" which would resolve https://github.com/kubernetes-sigs/kustomize/issues/947

Comment (laverya) [reply go.mod]: The full diff is here: https://github.com/go-yaml/yaml/compare/v3...laverya:v3.0.0-beta1
7 additional commits

**Comment by laverya** [comment]
>>all Unmarshal calls are now strict.
>
>was strict unmarshalling actually enabled by default without a toggle?
>this can be "quite" breaking for the k8s space if we move to v3.

Sounds like adding an explicit "UnmarshalUnsafe" function might be worthwhile in `go-yaml/yaml.v3`, then.

**Comment by neolit123** [comment]
> Sounds like adding an explicit "UnmarshalUnsafe" function might be worthwhile in go-yaml/yaml.v3, then.

negating API methods is breaking too. at least the change was part of a MAJOR, yet this leaves us in a space where we have to fork v3 if we want to move to it.

**Comment by laverya** [comment]
> negating API methods is breaking too. at least the change was part of a MAJOR, yet this leaves us in a space where we have to fork v3 if we want to move to it.

I meant something more along the lines of "add the negation method to the upstream, and use it to keep `kubernetes-sigs/yaml`'s methods the same". And since this is already based on a fork, what's one more change, right? 😆

**Comment by laverya** [comment]
@neolit123 unmarshal behavior with duplicate keys should be restored now through the use of `yaml3.UnmarshalPermissive`

**Comment by neolit123** [comment]
thanks for the update @laverya ,
one of the reasons we moved this yaml package under sigs.k8s.io as a fork is to not use third party/private dependency in a very core area of the project.

with this PR we are again introducing a private repository as a dependency.
but at this point will defer to the approvers whether we should proceed with the PR.

**Comment by laverya** [comment]
> one of the reasons we moved this yaml package under sigs.k8s.io as a fork is to not use third party/private dependency in a very core area of the project.
>
> with this PR we are again introducing a private repository as a dependency.

That makes perfect sense, and honestly I would not want my fork as a key dependency either 😆

I'd initially intended to wait for the upstream to merge things before making this PR, but as time goes on that's looking more and more like it won't happen without some sort of external impetus - and this could _be_ that impetus. Postponing a merge until we can switch back to `go-yaml/yaml` would make sense to me, but I think the rest of the code could still be reviewed today.

**Comment by fejta** [comment]
How would we feel about creating a v3 branch of this repo that uses (the patched) yaml.v3? 

This will downstream repos to safely start opting into the new behavior by changing their import path

**Comment by laverya** [comment]
I'd be happy with that! Either with a 'laverya' dependency or something actually owned by kubernetes

**Comment by thaJeztah** [comment]
I opened https://github.com/kubernetes-sigs/yaml/pull/29 to update the v2 version to v2.2.7

**Comment by horacimacias** [comment]
is there any chance in reopening this or creating a new PR to have this done?
I'm being hit by yaml 1.1 vs 1.2 differences and it sounds like this migration to yaml.v3 should help if not completely fix the issues.

**Comment by liggitt** [comment]
> is there any chance in reopening this or creating a new PR to have this done? I'm being hit by yaml 1.1 vs 1.2 differences and it sounds like this migration to yaml.v3 should help if not completely fix the issues.

https://github.com/kubernetes-sigs/yaml/pull/76 is plan to update this repo to use yaml.v3 while preserving compatibility with existing formatting

That PR is pending on some remaining comments - see https://github.com/kubernetes-sigs/yaml/pull/76#issuecomment-1351475437

**Pull request closed without merging**